runtime/neumann/src/fee_handler.rs

```diff
@@ -0,0 +1,163 @@
+		}
+	}
+}
+pub type CallOf<T> = <T as frame_system::Config>::Call;
+type NegativeImbalanceOf<C, T> =
+	<C as Currency<<T as frame_system::Config>::AccountId>>::NegativeImbalance;
+
+pub struct DuplicateCurrencyAdapter<MC, C, OU, TA, FCP>(PhantomData<(MC, C, OU, TA, FCP)>);
+
+impl<T, MC, C, OU, TA, FCP> OnChargeTransaction<T> for DuplicateCurrencyAdapter<MC, C, OU, TA, FCP>
+where
+	T: pallet_transaction_payment::Config,
+	C: Currency<<T as frame_system::Config>::AccountId>,
+	C::Balance: From<MC::Balance>,
+	C::PositiveImbalance: Imbalance<
+		<C as Currency<<T as frame_system::Config>::AccountId>>::Balance,
+		Opposite = C::NegativeImbalance,
+	>,
+	C::NegativeImbalance: Imbalance<
+		<C as Currency<<T as frame_system::Config>::AccountId>>::Balance,
+		Opposite = C::PositiveImbalance,
+	>,
+	MC::CurrencyId: From<TokenId>,
+	MC::Balance: From<C::Balance>,
+	MC: MultiCurrency<<T as frame_system::Config>::AccountId>,
+	OU: OnUnbalanced<NegativeImbalanceOf<C, T>>,
+	TA: Get<T::AccountId>,
+	FCP: CallParser<CallOf<T>>,
+{
+	type LiquidityInfo = Option<NegativeImbalanceOf<C, T>>;
+	type Balance = <C as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	/// Withdraw the prxedicted fee from the transaction origin.
+	///
+	/// Note: The `fee` already includes the `tip`.
+	fn withdraw_fee(
+		who: &T::AccountId,
+		call: &T::Call,
+		_info: &DispatchInfoOf<T::Call>,
+		fee: Self::Balance,
+		tip: Self::Balance,
+	) -> Result<Self::LiquidityInfo, TransactionValidityError> {
+		if fee.is_zero() {
+			return Ok(None)
+		}
+
+		// cross it
+		let call_information = FCP::fee_information(call.clone());
+
+		// TODO: when no info or we don't allow that token throw error
+
+		if call_information.token_id == NATIVE_TOKEN_ID {
+			let withdraw_reason = if tip.is_zero() {
+				WithdrawReasons::TRANSACTION_PAYMENT
+			} else {
+				WithdrawReasons::TRANSACTION_PAYMENT | WithdrawReasons::TIP
+			};
+
+			match C::withdraw(who, fee, withdraw_reason, ExistenceRequirement::KeepAlive) {
+				Ok(imbalance) => Ok(Some(imbalance)),
+				Err(_) => Err(InvalidTransaction::Payment.into()),
+			}
+		} else {
+			let currency_id = call_information.token_id.into();
+			// orml_tokens doesn't provide withdraw that allows setting existence so prevent
+			// withdraw if they don't have enough to avoid reaping
+			MC::ensure_can_withdraw(
+				currency_id,
+				who,
+				fee.saturating_add(MC::minimum_balance(currency_id).into()).into(),
+			)
+			.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			MC::withdraw(currency_id, who, fee.into())
+				.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			MC::deposit(currency_id, &TA::get(), fee.into()) // treasury account
+				.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			// TODO: Fire event for deposit
+
+			// We dealt with imbalance here so don't let `correct_and_deposit_fee` do it
+			Ok(None)
```

Comment:
Prevent `correct_and_deposit_fee` from doing anything when tokens are foreign.

Reply:
aren't you already doing that at the end of `ForeignFeeProcessor.withdraw_fee`?

Reply:
`ForeignFeeProcessor.withdraw_fee` returns an empty tuple and not an option.  We could change that, although I think I prefer the explicitness of this given how important returning `None` is.